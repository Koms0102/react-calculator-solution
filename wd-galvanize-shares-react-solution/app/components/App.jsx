import DeleteDialog from './DeleteDialog';
import LoginDialog from './LoginDialog';
import React from 'react';
import SideNav from './SideNav';
import Snackbar from 'material-ui/Snackbar';
import TopNav from './TopNav';
import axios from 'axios';

const App = React.createClass({
  getInitialState() {
    return {
      deleteDialog: {
        open: false,
        post: null
      },

      editing: [],

      loginDialog: {
        open: false
      },

      posts: [],

      sideNav: {
        open: false
      },

      snackbar: {
        message: '',
        open: false
      },

      user: null
    };
  },

  componentWillMount() {
    axios.get('/api/posts')
      .then((res) => {
        const nextPosts = res.data;

        this.setState({ posts: nextPosts });
      })
      .catch(() => {
        const nextSnackbar = {
          message: 'Unable to retrieve posts',
          open: true
        };

        this.setState({ snackbar: nextSnackbar });
      });
  },

  handleRequestCloseSnackbar() {
    const nextSnackbar = { message: '', open: false };

    this.setState({ snackbar: nextSnackbar });
  },

  closeDeleteDialog() {
    const nextDeleteDialog = { open: false, post: null };

    this.setState({ deleteDialog: nextDeleteDialog });
  },

  closeLoginDialog() {
    const nextLoginDialog = { open: false };

    this.setState({ loginDialog: nextLoginDialog });
  },

  createPost() {
    const post = {
      submitter: 'stanleypaddles',
      title: '',
      topic: this.props.params.topic || '',
      url: '',
      votes: Infinity
    };

    const nextEditing = this.state.editing.concat(post);
    const nextPosts = this.state.posts.concat(post);

    this.setState({ editing: nextEditing, posts: nextPosts });
  },

  deletePost(post) {
    const nextPosts = this.state.posts.filter((element) => post !== element);
    const nextDeleteDialog = { open: false, post: null };

    this.setState({ deleteDialog: nextDeleteDialog, posts: nextPosts });
  },

  incrementVotes(post, value) {
    if (!this.state.user) {
      const nextLoginDialog = { open: true };

      return this.setState({ loginDialog: nextLoginDialog });
    }

    const nextPosts = this.state.posts.map((element) => {
      if (post !== element) {
        return element;
      }

      const nextVotes = post.votes + value;
      const nextPost = Object.assign({}, post, { votes: nextVotes });

      return nextPost;
    });

    this.setState({ posts: nextPosts });
  },

  login(user) {
    const nextLoginDialog = { open: false };
    const nextSnackbar = { message: 'Logged in', open: true };
    const nextUser = user.username;

    this.setState({
      loginDialog: nextLoginDialog,
      snackbar: nextSnackbar,
      user: nextUser
    });

    return true;
  },

  logout() {
    const nextLoginDialog = { open: false };
    const nextSideNav = { open: false };
    const nextSnackbar = { message: 'Logged out', open: true };
    const nextUser = null;

    this.setState({
      loginDialog: nextLoginDialog,
      sideNav: nextSideNav,
      snackbar: nextSnackbar,
      user: nextUser
    });
  },

  openSideNav() {
    const nextSideNav = { open: true };

    this.setState({ sideNav: nextSideNav });
  },

  openLoginDialog() {
    const nextLoginDialog = { open: true };
    const nextSideNav = { open: false };

    this.setState({ loginDialog: nextLoginDialog, sideNav: nextSideNav });
  },

  openDeleteDialog(post) {
    const nextDeleteDialog = { open: true, post };

    this.setState({ deleteDialog: nextDeleteDialog });
  },

  startEditingPost(post) {
    const nextEditing = this.state.editing.concat(post);

    this.setState({ editing: nextEditing });
  },

  stopEditingPost(post) {
    const nextEditing = this.state.editing.filter((el) => post !== el);

    let nextPosts = this.state.posts;

    if (!Number.isFinite(post.votes)) {
      nextPosts = this.state.posts.filter((el) => post !== el);
    }

    this.setState({ editing: nextEditing, posts: nextPosts });
  },

  toggleSideNav(open) {
    const nextSideNav = { open };

    this.setState({ sideNav: nextSideNav });
  },

  updatePost(post, nextPost) {
    axios.post('/api/posts', nextPost)
      .then((res) => {
        const nextEditing = this.state.editing.filter((el) => post !== el);

        const nextPosts = this.state.posts.map((element) => {
          if (post !== element) {
            return element;
          }

          return res.data;
        });

        this.setState({ editing: nextEditing, posts: nextPosts });
      })
      .catch(() => {
        const nextSnackbar = {
          message: 'Unable to save post',
          open: true
        };

        this.setState({ snackbar: nextSnackbar });
      });
  },

  getChildrenProps() {
    const props = {
      '/': {
        editing: this.state.editing,
        incrementVotes: this.incrementVotes,
        openDeleteDialog: this.openDeleteDialog,
        params: this.props.params,
        posts: this.state.posts,
        startEditingPost: this.startEditingPost,
        stopEditingPost: this.stopEditingPost,
        updatePost: this.updatePost
      }
    };

    props['/topics/:topic'] = props['/'];

    const matchPath = this.props.routes.reduce((path, route) => {
      return path + (route.path || '');
    }, '');

    return props[matchPath];
  },

  render() {
    return <div>
      <DeleteDialog
        closeDeleteDialog={this.closeDeleteDialog}
        deleteDialog={this.state.deleteDialog}
        deletePost={this.deletePost}
      />

      <LoginDialog
        closeLoginDialog={this.closeLoginDialog}
        login={this.login}
        loginDialog={this.state.loginDialog}
      />

      <SideNav
        createPost={this.createPost}
        logout={this.logout}
        openLoginDialog={this.openLoginDialog}
        sideNav={this.state.sideNav}
        toggleSideNav={this.toggleSideNav}
        user={this.state.user}
      />

      <Snackbar
        autoHideDuration={3000}
        message={this.state.snackbar.message}
        onRequestClose={this.handleRequestCloseSnackbar}
        open={this.state.snackbar.open}
      />

      <TopNav
        createPost={this.createPost}
        logout={this.logout}
        openLoginDialog={this.openLoginDialog}
        openSideNav={this.openSideNav}
        user={this.state.user}
      />

      {React.cloneElement(
        this.props.children,
        this.getChildrenProps()
      )}
    </div>;
  }
});

export default App;
