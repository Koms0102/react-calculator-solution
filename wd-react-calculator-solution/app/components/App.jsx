import React from 'react';

const ENTER_KEY = 13;
const ESCAPE_KEY = 27;

const App = React.createClass({
  getInitialState() {
    return { screen: '' };
  },

  handleChange(event) {
    const nextState = { screen: event.target.value };

    this.setState(nextState);
  },

  handleClickClear() {
    this.setState({ screen: '' });
  },

  handleClickEquals() {
    if (this.state.screen === 'Error') {
      return;
    }

    const regexp = /^(\-?\d+)(x|\/|\+|\-)(\-?\d+)$/;

    const matches = this.state.screen.match(regexp);

    if (matches === null) {
      this.setState({ screen: 'Error' });

      return;
    }

    const operand1 = parseInt(matches[1]);
    const operand2 = parseInt(matches[3]);
    const operator = matches[2];
    const evaluate = {
      x: (aa, bb) => aa * bb,
      '+': (aa, bb) => aa + bb,
      '-': (aa, bb) => aa - bb,
      '*': (aa, bb) => aa * bb,
      '/': (aa, bb) => aa / bb
    };
    const total = evaluate[operator](operand1, operand2);
    const nextScreen = total.toString();

    this.setState({ screen: nextScreen });
  },

  handleClickOp(event) {
    if (this.state.screen === 'Error') {
      return;
    }

    const nextScreen = this.state.screen + event.target.textContent;

    this.setState({ screen: nextScreen });
  },

  handleKeyDown(event) {
    if (event.which === ENTER_KEY) {
      this.handleClickEquals();
    }
    else if (event.which === ESCAPE_KEY) {
      event.preventDefault();
      this.handleClickClear();
    }
  },

  render() {
    return <div id="calculator">
      <div id="screen-container">
        <input
          id="screen"
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
          value={this.state.screen}
        />
      </div>

      <div id="button-container">
        <div className="buttons">
          <span
            className="operator"
            id="cancel"
            onClick={this.handleClickClear}
          >
            C
          </span>
          <span
            className="operator"
            onClick={this.handleClickOp}
          >
            /
          </span>
          <span
            className="operator"
            onClick={this.handleClickOp}
          >
            x
          </span>
          <span onClick={this.handleClickOp}>7</span>
          <span onClick={this.handleClickOp}>8</span>
          <span onClick={this.handleClickOp}>9</span>
          <span
            className="operator"
            onClick={this.handleClickOp}
          >
            -
          </span>
          <span onClick={this.handleClickOp}>4</span>
          <span onClick={this.handleClickOp}>5</span>
          <span onClick={this.handleClickOp}>6</span>
          <span
            className="operator"
            onClick={this.handleClickOp}
          >
            +
          </span>
          <span onClick={this.handleClickOp}>1</span>
          <span onClick={this.handleClickOp}>2</span>
          <span onClick={this.handleClickOp}>3</span>
          <span
            className="operator"
            id="calc"
            onClick={this.handleClickEquals}
          >
            =
          </span>
          <div className="l-row">
            <span id="zero" onClick={this.handleClickOp}>0</span>
          </div>
        </div>
      </div>
    </div>;
  }
});

export default App;
