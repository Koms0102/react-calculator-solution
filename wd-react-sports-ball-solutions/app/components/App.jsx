import React from 'react';

const App = React.createClass({
  getInitialState() {
    return { place: '' };
  },

  handleChange(event) {
    const nextState = { place: event.target.value };

    this.setState(nextState);
  },

  render() {
    const place = this.state.place.trim().toLowerCase();

    let message;

    if (place === 'seattle') {
      message = <ul className="no-bullets">
        <li style={{ margin: '10px' }}>
          <a href="http://www.soundersfc.com/">Go Sounders!</a>
        </li>
        <li style={{ margin: '10px' }}>
          <img
            src="https://upload.wikimedia.org/wikipedia/en/2/27/Seattle_Sounders_FC.svg"
            width="150"
          />
        </li>
      </ul>;
    }
    else if (place === 'portland') {
      message = <ul className="no-bullets">
        <li style={{ margin: '10px' }}>
          <a href="http://www.trailblazers.com/">Go Trail Blazers!</a>
        </li>
        <li style={{ margin: '10px' }}>
          <img
            src="https://upload.wikimedia.org/wikipedia/en/0/06/Portland_Trail_Blazers_alternate_logo.svg"
            width="150"
          />
        </li>
      </ul>;
    }
    else if (place === '') {
      message = <p>Go ahead, type in a city.</p>;
    }
    else {
      message = <p>{this.state.place}? Never heard of it.</p>;
    }

    return <div>
      <h1>
        {"What's your favorite city?"}
      </h1>
      <input
        onChange={this.handleChange}
        type="text"
        value={this.state.place}
      />
      <div>{message}</div>
    </div>;
  }
});

export default App;
