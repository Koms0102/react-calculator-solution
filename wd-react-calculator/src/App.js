import React from 'react';
import './App.css';
import Screen from './Screen/Screen'
import Buttons from './Buttons/Buttons'
import KeyPressDetector from './KeyPressDetector/KeyPressDetector'

class App extends React.Component {
  constructor() {
    super();
    this.state = {
      screen: ''
    }

    this.handleClickOperator = this.handleClickOperator.bind(this);
    this.handleClickNumber = this.handleClickNumber.bind(this);
    this.clearScreen = this.clearScreen.bind(this);
    this.displayError = this.displayError.bind(this);
    this.handleClickEquals =this.handleClickEquals.bind(this);
    this.handleKeyPress =this.handleKeyPress.bind(this);
  }

  handleClickNumber(num) {
    if(this.state.screen === 'Error') return;
    this.setState({screen: this.state.screen + num});
  }

  handleClickOperator(op) {
    if(this.state.screen === 'Error') return;
    this.setState({screen: this.state.screen + op});
  }

  clearScreen() {
    this.setState({screen: ''});
  }

  handleClickEquals() {
    if(this.state.screen === 'Error') return;

    const regexp = /^(\-?\d+)(x|\-|\+|\/)(\-?\d+)$/;

    const evaluate = {
      'x': (a, b) => a * b, 
      '/': (a, b) => a / b,
      '-': (a, b) => a - b,
      '+': (a, b) => a + b
    }

    const matches = this.state.screen.match(regexp); 

    if(matches === null) {
      this.displayError();
      return;
    }

    const number1 = parseInt(matches[1]);
    const number2 = parseInt(matches[3]);
    const operator = matches[2];

    const result = evaluate[operator](number1, number2);

    this.setState({screen: result})

  }

  displayError() {
    this.setState({screen: 'Error'});
  }

  handleKeyPress(key) {
    const ops = ['+', '-', 'x', '/'];
    if(ops.includes(key)) this.handleClickOperator(key)
    else if(key === 'Enter') this.handleClickEquals()
    else if(!isNaN(Number(key))) this.handleClickNumber(key)

  }

  render() {
    return (
      <KeyPressDetector
        handleKeyPress={this.handleKeyPress}
      >
        <div id="calculator">
          <Screen displayValue={this.state.screen} />
          <Buttons 
            handleClickNumber={this.handleClickNumber}
            handleClickOperator={this.handleClickOperator}
            handleClickEquals={this.handleClickEquals}
            handleClickClear={this.clearScreen}
          />
        </div>
      </KeyPressDetector>
    );
  }
}

export default App;
