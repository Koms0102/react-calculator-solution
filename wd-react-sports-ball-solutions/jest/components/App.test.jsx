import App from '../../app/components/App';
import React from 'react';
import renderer from 'react-test-renderer';

describe('<App />', () => {
  it('renders', () => {
    const component = renderer.create(<App />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with some place', () => {
    const component = renderer.create(<App />);
    const event = { target: { value: 's' }};

    let tree = component.toJSON();

    tree.children[1].props.onChange(event);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with seattle', () => {
    const component = renderer.create(<App />);
    const event = { target: { value: 'seattle' }};

    let tree = component.toJSON();

    tree.children[1].props.onChange(event);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with SeAtTlE', () => {
    const component = renderer.create(<App />);
    const event = { target: { value: 'SeAtTlE' }};

    let tree = component.toJSON();

    tree.children[1].props.onChange(event);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with portland', () => {
    const component = renderer.create(<App />);
    const event = { target: { value: 'portland' }};

    let tree = component.toJSON();

    tree.children[1].props.onChange(event);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
