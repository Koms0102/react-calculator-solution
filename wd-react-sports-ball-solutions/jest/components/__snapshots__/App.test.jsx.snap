exports[`<App /> renders 1`] = `
<div>
  <h1>
    What\'s your favorite city?
  </h1>
  <input
    onChange={[Function bound handleChange]}
    type="text"
    value="" />
  <div>
    <p>
      Go ahead, type in a city.
    </p>
  </div>
</div>
`;

exports[`<App /> renders with SeAtTlE 1`] = `
<div>
  <h1>
    What\'s your favorite city?
  </h1>
  <input
    onChange={[Function bound handleChange]}
    type="text"
    value="SeAtTlE" />
  <div>
    <ul
      className="no-bullets">
      <li
        style={
          Object {
            "margin": "10px"
          }
        }>
        <a
          href="http://www.soundersfc.com/">
          Go Sounders!
        </a>
      </li>
      <li
        style={
          Object {
            "margin": "10px"
          }
        }>
        <img
          src="https://upload.wikimedia.org/wikipedia/en/2/27/Seattle_Sounders_FC.svg"
          width="150" />
      </li>
    </ul>
  </div>
</div>
`;

exports[`<App /> renders with portland 1`] = `
<div>
  <h1>
    What\'s your favorite city?
  </h1>
  <input
    onChange={[Function bound handleChange]}
    type="text"
    value="portland" />
  <div>
    <ul
      className="no-bullets">
      <li
        style={
          Object {
            "margin": "10px"
          }
        }>
        <a
          href="http://www.trailblazers.com/">
          Go Trail Blazers!
        </a>
      </li>
      <li
        style={
          Object {
            "margin": "10px"
          }
        }>
        <img
          src="https://upload.wikimedia.org/wikipedia/en/0/06/Portland_Trail_Blazers_alternate_logo.svg"
          width="150" />
      </li>
    </ul>
  </div>
</div>
`;

exports[`<App /> renders with seattle 1`] = `
<div>
  <h1>
    What\'s your favorite city?
  </h1>
  <input
    onChange={[Function bound handleChange]}
    type="text"
    value="seattle" />
  <div>
    <ul
      className="no-bullets">
      <li
        style={
          Object {
            "margin": "10px"
          }
        }>
        <a
          href="http://www.soundersfc.com/">
          Go Sounders!
        </a>
      </li>
      <li
        style={
          Object {
            "margin": "10px"
          }
        }>
        <img
          src="https://upload.wikimedia.org/wikipedia/en/2/27/Seattle_Sounders_FC.svg"
          width="150" />
      </li>
    </ul>
  </div>
</div>
`;

exports[`<App /> renders with some place 1`] = `
<div>
  <h1>
    What\'s your favorite city?
  </h1>
  <input
    onChange={[Function bound handleChange]}
    type="text"
    value="s" />
  <div>
    <p>
      s
      ? Never heard of it.
    </p>
  </div>
</div>
`;
